v4/export/dump.go

```diff
@@ -196,193 +188,386 @@ func (d *Dumper) Dump() (err error) {
+		query := buildSelectQuery(db, tbl, selectField, buildWhereCondition(conf, where), orderByClause)
+		if len(nullValueCondition) > 0 {
+			nullValueCondition = ""
+		}
+		ctxDone := sendDataIRToChan(ctx, newTableData(query, selectLen), ir)
+		if ctxDone {
+			break
+		}
+		cutoff = nextCutOff
 	}
 	return nil
 }
 
-func dumpTable(ctx context.Context, conf *Config, db *sql.Conn, dbName string, table *TableInfo, writer Writer) ([]TableDataIR, error) {
-	tableName := table.Name
-	if !conf.NoSchemas {
-		if table.Type == TableTypeView {
-			viewName := table.Name
-			createTableSQL, createViewSQL, err := ShowCreateView(db, dbName, viewName)
-			if err != nil {
-				return nil, err
-			}
-			return nil, writer.WriteViewMeta(ctx, dbName, viewName, createTableSQL, createViewSQL)
+func sendDataIRToChan(ctx context.Context, ir TableDataIR, irChan chan<- TableDataIR) (ctxDone bool) {
+	for {
+		select {
+		case <-ctx.Done():
+			return true
+		case irChan <- ir:
+			return false
+		default:
+			sleepTime := 1 * time.Second
+			time.Sleep(sleepTime)
+			log.Warn("write channel is full, waiting writer to consume IR",
+				zap.Int("channel size", defaultDumpThreads),
+				zap.Stringer("sleep time", sleepTime))
 		}
-		createTableSQL, err := ShowCreateTable(db, dbName, tableName)
+	}
+}
+
+var z = &big.Int{}
+
+func (d *Dumper) selectMinAndMaxIntValue(conn *sql.Conn, db, tbl, field string) (*big.Int, *big.Int, error) {
+	ctx, conf := d.ctx, d.conf
+	query := fmt.Sprintf("SELECT MIN(`%s`),MAX(`%s`) FROM `%s`.`%s`",
+		escapeString(field), escapeString(field), escapeString(db), escapeString(tbl))
+	if conf.Where != "" {
+		query = fmt.Sprintf("%s WHERE %s", query, conf.Where)
+	}
+	log.Debug("split chunks", zap.String("query", query))
+
+	var smin sql.NullString
+	var smax sql.NullString
+	row := conn.QueryRowContext(ctx, query)
+	err := row.Scan(&smin, &smax)
+	if err != nil {
+		log.Error("split chunks - get max min failed", zap.String("query", query), zap.Error(err))
+		return z, z, err
+	}
+	if !smax.Valid || !smin.Valid {
+		// found no data
+		log.Warn("no data to dump", zap.String("schema", db), zap.String("table", tbl))
+		return z, z, nil
+	}
+
+	max := new(big.Int)
+	min := new(big.Int)
+	var ok bool
+	if max, ok = max.SetString(smax.String, 10); !ok {
+		return z, z, errors.Errorf("fail to convert max value %s in query %s", smax.String, query)
+	}
+	if min, ok = min.SetString(smin.String, 10); !ok {
+		return z, z, errors.Errorf("fail to convert min value %s in query %s", smin.String, query)
+	}
+	return min, max, nil
+}
+
+func (d *Dumper) concurrentDumpTiDBTables(conn *sql.Conn, meta TableMeta, ir chan<- TableDataIR) error {
+	ctx, conf := d.ctx, d.conf
+	db, tbl := meta.DatabaseName(), meta.TableName()
+
+	handleColNames, handleVals, err := selectTiDBTableSample(db, tbl, conn, meta)
+	if err != nil {
+		return err
+	}
+	if len(handleVals) == 0 {
+		return nil
```

Comment:
Should we return an error or some instructions here to help `sequentialDumpTable` or `concurrentDumpTable` using the unique index? This table will directly be skipped if we `return nil` here.

Reply:
If the `handleVals` is empty, there is no data in this table.